lib/pure/asyncdispatch.nim

```diff
@@ -1559,8 +1559,8 @@ proc drain*(timeout = 500) =
   ## Waits for completion events and processes them. Raises ``ValueError``
   ## if there are no pending operations. In contrast to ``poll`` this
   ## processes as many events as are available.
-  if runOnce(timeout):
-    while hasPendingOperations() and runOnce(0): discard
+  if runOnce(timeout) or hasPendingOperations():
```

Comment:
This is the most controversial change of this PR. I would like to get the opinions of @dom96 and @zevv 

My understanding is that there is an implicit Post Condition for `drain()`:
- `hasPendingOperations()` should **always** return `false` after `drain()` is called.

My understanding of `runOnce()` is that the returned boolean represents if some *work* was done (i.e. callbacks were executed). 

It's possible for `runOnce()` to be called, return false, and `hasPendingOperations()` still be true.
If that case happens, the while loop never runs, and `drain()` does not fulfill it's Post Condition.

Funny thing: I don't think this case ever happens on unix platforms, because of the way the ioselector implementations are scheduled. But, on Windows, it's entirely possible for this to happen, depending on how the Windows Kernel decides to schedule the Select Events (particularly on slow environments like a VM). I believe this is why the CI windows builds were being flaky.

Reply:
What you say makes sense, but to be honest, if it was up to me I would get rid of ``drain``. I'm probably missing something, but I have never needed to use it nor do I understand its use case.

Reply:
`drain` is very useful for testing lol (I use it in the tests I added to the test suite)

It provides a nice way to test parts of asyncdispatch in a deterministic way that doesn't involve hacks like throwing exceptions to stop the run loop.

I could see a valid situation where `drain` could be used to assist with implementing a *graceful shutdown*.
For example: You have a seperate thread with an async-loop that receives a shutdown signal. You might want to make sure all async operations have successfully completed before you destroy the thread.

Also, `drain` has been here for a long time and I don't feel comfortable removing it, for fear of breaking code in the wild.

Reply:
So if I understand correctly, your use case for ``drain`` is to run the event loop until a test finishes? That sounds like precisely what ``waitFor`` is for...

> You might want to make sure all async operations have successfully completed before you destroy the thread. (I'm grasping a bit here. I'm sure there are many other ways to do this...)

I suppose, but you could also just use `waitFor` here.

> Mainly, drain has been here for a long time and I don't feel comfortable removing it, for fear of breaking code in the wild.

I don't think you should remove it in this PR. I'm just expressing my frustration about this function.

Reply:
`WaitFor` is not the same as all:
Promises can be interpreted as a dependency graph.
`WaitFor` Only works if you assume your promise graph is a tree and you hold the root node. 
What if the set of promises is not a tree? or what if you don't have access to the root promise?
What if there are several promise trees and you want to process all of them?

An http server is a good example. There is no single `promise` to wait on. The server receives requests and starts an async Handler for each request.
Now imagine there is an error case. Maybe some *async logger* fails because hard drive is full.
You might be able to catch the exception in that one single request, by using `waitFor` on the promise for that particular request.

**But**, there are bunch of outstanding requests in the async event loop (maybe some of them have already written to the log, and are just waiting to send their result).
What is there to `waitFor` in that case? How do we get access to all of those requests promises?
A much easier solution is to get the global PDispatcher, and then `drain` the asyncDispatch queue, allowing each request to individually send it's own response and finish, then gracefully shutdown.

I find it odd that you feel this way, considering some of the stuff you had to do in your custom event loop in httpbeast.
You solve this in httpbeast by reaching into the PDispatcher internals and forcing the queue to continue. That is kind of invasive imo. `Drain` allows a more modular approach.

Reply:
> WaitFor Only works if you assume your promise graph is a tree and you hold the root node.

That's a fair point, but I have difficulty imagining such a use case. Even if it exists, it's a niche use case and as such doesn't deserve it's own ``drain`` procedure.

> You solve this in httpbeast by reaching into the PDispatcher internals and forcing the queue to continue. That is kind of invasive imo.

That's not correct at all. The `poll` proc is a part of the dispatcher's API, just like it is for the selectors module. The ``waitFor`` and ``runForever`` are just convenience procedures.

The reason I don't use asyncdispatcher's event loop is because I wanted to control every aspect of httpbeast for the best performance possible, being able to disable asyncdispatch processing is pretty powerful and if I ran it on asyncdispatch I don't think I'd be able to implement parallelism.

Reply:
> That's not correct at all. The poll proc is a part of the dispatcher's API, just like it is for the selectors module. The waitFor and runForever are just convenience procedures.

That's a fair point, but if you look at it that way, `drain` is just another convenience proc, that you can choose to use if it makes sense for you as a user of the library. I don't see how it hurts anything.

> The reason I don't use asyncdispatcher's event loop is because I wanted to control every aspect of httpbeast for the best performance possible, being able to disable asyncdispatch processing is pretty powerful and if I ran it on asyncdispatch I don't think I'd be able to implement parallelism.

Fair point, but this is exactly the problem I'm tackling. My ultimate goal with all of these PR's is to make asyncdispatch better able to deal with parallelism.

As soon as parallelism is introduced, (I.e. multiple asyncdispatch loops on different threads),
The dependency graph becomes much more complicated, and these kinds of convenience methods become much more useful.